chemprop/models/ffn.py

```diff
@@ -0,0 +1,250 @@
+        """
+        :param features_size: Dimensionality of input features.
+        :param hidden_size: Dimensionality of hidden layers.
+        :param num_layers: Number of layers in FFN.
+        :param output_size: The size of output.
+        :param dropout: Dropout probability.
+        :param activation: Activation function.
+        :param constraint: Whether to apply constraint to output.
+        :param ffn_type: The type of target (atom or bond).
+        """
+        super(FFNAtten, self).__init__()
+
+        if constraint:
+            self.ffn = DenseLayers(first_linear_dim=features_size,
+                                   hidden_size=hidden_size,
+                                   num_layers=num_layers,
+                                   output_size=hidden_size,
+                                   dropout=dropout,
+                                   activation=activation)
+            self.ffn_readout = DenseLayers(first_linear_dim=hidden_size,
+                                           hidden_size=hidden_size,
+                                           num_layers=1,
+                                           output_size=output_size,
+                                           dropout=dropout,
+                                           activation=activation)
+            self.weights_readout = DenseLayers(first_linear_dim=hidden_size,
+                                               hidden_size=hidden_size,
+                                               output_size=1,
+                                               num_layers=2,
+                                               dropout=dropout,
+                                               activation=activation)
+        else:
+            self.ffn_readout = DenseLayers(first_linear_dim=features_size,
+                                           hidden_size=hidden_size,
+                                           num_layers=num_layers,
+                                           output_size=output_size,
+                                           dropout=dropout,
+                                           activation=activation)
+        self.constraint = constraint
+        self.ffn_type = ffn_type
+
+    def forward(self, input: Tuple[torch.tensor, List, torch.tensor, List, torch.tensor], constraints: torch.tensor) -> torch.tensor:
+        """
+        Runs the :class:`FFNAtten` on input.
+        :param input: A tuple of atom and bond informations of each molecule.
+        :return: The output of the :class:`FFNAtten`, a PyTorch tensor containing a list of property predictions.
+        """
+        a_hidden, a_scope, b_hidden, b_scope, b2br = input
+
+        if self.ffn_type == 'atom':
+            hidden = a_hidden
+            scope = a_scope
+
+        elif self.ffn_type == 'bond':
+            forward_bond = b_hidden[b2br[:, 0]]
+            backward_bond = b_hidden[b2br[:, 1]]
+            hidden = torch.cat([forward_bond, backward_bond], dim=1)
```

Comment:
This looks order dependent. I don't think we'd want the output of the bond prediction to depend on the arbitrary choice of which direction is considered forward. Chances are the model will converge for them to be similar, but we'd really like for them to be order invariant. Ideally, this would even help the model train faster because you wouldn't have to learn the weights for forward and reverse separately.

I'm not sure how we'd do that though. Maybe there's a way we can force the weights to be symmetrical across the two halves of the input. That would be the best way. We could also use some kind of aggregation function to add them together, but then we'd lose some information. I'll think on this and see if I can come up with anyting.